### ISSUE TITLE
PR #1292 (Boat.Type enum extension) causes crashing in real-world use

### ISSUE THREAD

**Issue by Tslat** [neoforged/NeoForge#1567]
So it appears that although the PR itself (#1292) included a test case, it got lucky in that its test case involves registering everything simultaneously in the same class.

This means that for mods that do registration in their own registration classes (I.E., most mods), the test case doesn't match real-world conditions.
This means that trying to add a custom boat type appears to cause a registration death loop before the game even gets going.

see here:

```
Caused by: java.lang.NullPointerException: Cannot invoke "net.neoforged.bus.api.IEventBus.addListener(java.util.function.Consumer)" because "bus" is null
	at TRANSFORMER/neoforge@21.1.62/net.neoforged.neoforge.registries.DeferredRegister.register(DeferredRegister.java:319) ~[neoforge-21.1.62-minecraft-merged.jar%23184!/:?] {re:classloading}
	at TRANSFORMER/aoa3@3.7.13/net.tslat.aoa3.common.registration.AoARegistries$RegistryHelper.<init>(AoARegistries.java:209) ~[%23185!/:?] {re:classloading}
	at TRANSFORMER/aoa3@3.7.13/net.tslat.aoa3.common.registration.AoARegistries$RegistryHelper.<init>(AoARegistries.java:203) ~[%23185!/:?] {re:classloading}
	at TRANSFORMER/aoa3@3.7.13/net.tslat.aoa3.common.registration.AoARegistries.<clinit>(AoARegistries.java:108) ~[%23185!/:?] {re:classloading}
	at TRANSFORMER/aoa3@3.7.13/net.tslat.aoa3.common.registration.block.AoABlocks.register(AoABlocks.java:785) ~[%23185!/:?] {re:classloading}
	at TRANSFORMER/aoa3@3.7.13/net.tslat.aoa3.common.registration.block.group.StoneBlockGroup.<init>(StoneBlockGroup.java:19) ~[%23185!/:?] {re:classloading}
	at TRANSFORMER/aoa3@3.7.13/net.tslat.aoa3.common.registration.block.AoABlocks.registerGroup(AoABlocks.java:793) ~[%23185!/:?] {re:classloading}
	at TRANSFORMER/aoa3@3.7.13/net.tslat.aoa3.common.registration.block.AoABlocks.<clinit>(AoABlocks.java:61) ~[%23185!/:?] {re:classloading}
	at TRANSFORMER/aoa3@3.7.13/net.tslat.aoa3.common.registration.entity.AoABoatTypes.<clinit>(AoABoatTypes.java:13) ~[%23185!/:?] {re:classloading}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.world.entity.vehicle.Boat$Type.<clinit>(Boat.java:931) ~[neoforge-21.1.62-minecraft-merged.jar%23183!/:?] {re:classloading,pl:runtime_enum_extender:B}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.world.item.Items.<clinit>(Items.java:890) ~[neoforge-21.1.62-minecraft-merged.jar%23183!/:?] {re:classloading}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.world.level.block.LightBlock.getShape(LightBlock.java:64) ~[neoforge-21.1.62-minecraft-merged.jar%23183!/:?] {re:classloading}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.world.level.block.state.BlockBehaviour$BlockStateBase.getShape(BlockBehaviour.java:724) ~[neoforge-21.1.62-minecraft-merged.jar%23183!/:?] {re:classloading}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.world.level.block.state.BlockBehaviour$BlockStateBase.getShape(BlockBehaviour.java:720) ~[neoforge-21.1.62-minecraft-merged.jar%23183!/:?] {re:classloading}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.world.level.block.state.BlockBehaviour.getCollisionShape(BlockBehaviour.java:366) ~[neoforge-21.1.62-minecraft-merged.jar%23183!/:?] {re:classloading,pl:accesstransformer:B}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.world.level.block.state.BlockBehaviour$BlockStateBase$Cache.<init>(BlockBehaviour.java:997) ~[neoforge-21.1.62-minecraft-merged.jar%23183!/:?] {re:classloading}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.world.level.block.state.BlockBehaviour$BlockStateBase.initCache(BlockBehaviour.java:557) ~[neoforge-21.1.62-minecraft-merged.jar%23183!/:?] {re:classloading}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.world.level.block.Blocks.<clinit>(Blocks.java:7811) ~[neoforge-21.1.62-minecraft-merged.jar%23183!/:?] {re:classloading,pl:accesstransformer:B}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.world.level.block.FireBlock.bootStrap(FireBlock.java:350) ~[neoforge-21.1.62-minecraft-merged.jar%23183!/:?] {re:classloading,pl:accesstransformer:B}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.server.Bootstrap.bootStrap(Bootstrap.java:48) ~[neoforge-21.1.62-minecraft-merged.jar%23183!/:?] {re:classloading}
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.client.main.Main.lambda$main$0(Main.java:114) ~[neoforge-21.1.62-minecraft-merged.jar%23183!/:?] {re:classloading,pl:runtimedistcleaner:A}
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572) ~[?:?] {}
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317) ~[?:?] {}
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?] {}
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?] {}
	at java.base/java.lang.Thread.run(Thread.java:1583) ~[?:?] {}
```

Basically, the game is boostrapping vanilla Item Registration, which causes this chain:

Bootstrap -> Vanilla Item Registration -> Boat Type clinit -> enum extension init -> Mod Block Registration clinit -> crash

This is due to the enum extension constructor requiring registered objects in it:
![image](https://github.com/user-attachments/assets/a13a91ba-4476-4e1d-888a-3907967cf25a)

**Comment by Matyrobbrt** [comment]
This works as intended. You're relying on your mod constructor having been called and the bus assigned (see https://github.com/Tslat/Advent-Of-Ascension/blob/1.21/source%2Fcommon%2Fregistration%2FAoARegistries.java#L209) during clinit of your registration class. This is not a safe assumption in any context as that class may be initialized early through other means (such as mixins that inject very early).
The recommended way to register your DR to the bus is to do it in a entrypoint. In your case, instead of the line I linked above, you can simply toss all the DRs in a list and iterate and register all of them in your entrypoint.

**Comment by Tslat** [comment]
Ah can confirm, thanks
